src/lib/expansion/expansion-panel.ts

```diff
@@ -147,16 +161,41 @@ export class MatExpansionPanel extends CdkAccordionItem
   _bodyAnimation(event: AnimationEvent) {
     const classList = event.element.classList;
     const cssClass = 'mat-expanded';
-    const {phaseName, toState} = event;
+    const {phaseName, toState, fromState} = event;
 
     // Toggle the body's `overflow: hidden` class when closing starts or when expansion ends in
     // order to prevent the cases where switching too early would cause the animation to jump.
     // Note that we do it directly on the DOM element to avoid the slight delay that comes
     // with doing it via change detection.
-    if (phaseName === 'done' && toState === 'expanded') {
-      classList.add(cssClass);
-    } else if (phaseName === 'start' && toState === 'collapsed') {
-      classList.remove(cssClass);
+    if (phaseName === 'done') {
+      if (toState === 'expanded') {
+        classList.add(cssClass);
+        if (fromState !== 'void') {
+          this.afterExpanded.emit();
+        }
+        return;
```

Comment:
I'm not sure whether all of these returns are necessary. If the state is something different, it just won't go into the other conditions. Also can we break this up a bit since it's a fair bit of `if`-s in one place?